### PR TITLE
Compute pace update

### PR DESCRIPTION
**Pull request by jmgoff**
**Summary**

With the growing number of atomic cluster expansion (ACE) users, a LAMMPS compute is needed at this time that can evaluate the descriptors for various methods and workflows such as FitSNAP and ML-IAP. This update features a new LAMMPS compute, `compute pace`, that evaluates ACE descriptors. The style mirrors the compute for SNAP descriptors, `compute snap`, and provides users a familiar method to evaluate ACE descriptors in LAMMPS for data analysis, for model training, etc.

**Related Issue(s)**

N/A

**Author(s)**
(Alphabetical)
 
Aidan Thompson (Sandia National Laboratories)
Andrew Rohskopf (Sandia National Laboratories)
James Goff (Sandia National Laboratories) ( goffjimmy1@gmail.com ) 
Yury Lysogorskiy (Ruhr University for lammps-user-pace backend) 

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**
No problems with old inputs.

**Implementation Notes**

The calculation of ACE descriptors and descriptor gradients were validated numerically against those from the existing `pair_style pace` in LAMMPS. They have also been validated against other ACE software that is not in LAMMPS.

This initial implementation of `compute pace` requires minor changes to the `lammps-user-pace` library that is downloaded whenever the  ML-PACE D flag is turned on for compilation. The required changes to the `lammps-user-pace` library are essentially just allocating arrays to store ACE descriptors and gradients. Until such changes are made in the default `lamps-user-pace` repo at ICAMS, some errors may be thrown unless a [separate fork](https://github.com/jmgoff/lammps-user-pace-1) of the `lammps-user-pace` library is used. Future updates and integration with the (external ACE compute)[https://github.com/ICAMS/python-ace/issues/14] developed by @yury-lysogorskiy at ICAMS would resolve this as well.

This first `compute pace` is here to provide users with the ACE descriptor compute capabilities immediately and serves as the base framework to integrate [more standardized descriptor + gradient calculators](https://github.com/ICAMS/python-ace/blob/feature/linear-fit/lib/ace/ace-evaluator/src/extra/ace_calculator.cpp) in the future. 

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

Link to the original ICAMS PACE libraries and those including an updated descriptor + gradient calculator: https://github.com/ICAMS/lammps-user-pace and https://github.com/ICAMS/python-ace/issues/14
Link to modified lammps-user-pace: https://github.com/jmgoff/lammps-user-pace-1

### CODE REVIEW

File: src/ML-PACE/compute_pace.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+// clang-format off
+ComputeStyle(pace,ComputePACE);
+// clang-format on
+#else
+
+#ifndef LMP_COMPUTE_PACE_H
+#define LMP_COMPUTE_PACE_H
+
+#include "compute.h"
+#include "ace-evaluator/ace_c_basis.h"
```
Comment (akohlmey) [inline src/ML-PACE/compute_pace.h]: These includes should not be here but rather used in compute_pace.cpp. It should be sufficient to have forward declaration of class instances in the header. Including application/library header files in a style class header carry a high potential for namespace pollution and unexpected compilation failures. Please see the corresponding notes in the programming style guide. https://docs.lammps.org/Modify_style.html#include-files-varied

If simple forward declarations cannot be used, the setup needs to be "PIMPL-ified", ie. a struct is used with all specific definitions and pointers and then the style header only needs to contain a pointer to that struct (with a forward declaration again) and the struct is then instantiated in the constructor. Please see pair style pacs for an example.

**Review by akohlmey — CHANGES_REQUESTED** [review]
I've added a comment for a required change to the compute_pace.h file.
This code also seems to be dependent on a different version of the ACE library since our CI testing fails to compile with errors indicating missing definitions.  See, e.g. https://ci.lammps.org/job/dev/job/pull_requests/job/fedora/job/cmake_kokkos_mpi_openmp_clang_shared/7591/console
Thus the corresponding build files need to be updated to download a suitable version.

File: src/ML-PACE/compute_pace.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
+/* -*- c++ -*- ----------------------------------------------------------
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   https://www.lammps.org/, Sandia National Laboratories
+   Steve Plimpton, sjplimp@sandia.gov
```
Comment (akohlmey) [inline src/ML-PACE/compute_pace.h]: The contact email should be:
```
LAMMPS development team: developers@lammps.org
```
It should be possible to update this with `make fix-homepage`
Also, please run `make fix-whitespace` in the src folder and commit those changes as well to have consistent whitespace (no tabs, no trailing whitespace). If you want to go one step further, you could also run clang-format on both, the header and the implementation.

File: src/ML-PACE/compute_pace.h (hunk 1 above)
Comment (jmgoff) [reply src/ML-PACE/compute_pace.h]: Forward declarations are hard to use here so it has been "PIMPL-ified" instead in the most recent commit.

Comment (jmgoff) [reply src/ML-PACE/compute_pace.h]: We still need to update cmake build scripts and/or ACE library to fix the CI testing problem.  While we work on a more elegant solution for this, the non-standard ACE libraries can be used if the following D flags are specified during compilation:

-DPACELIB_URL= http://github.com/jmgoff/lammps-user-pace-1/archive/refs/tags/lammps-user-pace-compute.tar.gz
-DPACELIB_MD5=0616251507dffe83dbd38681d101aee8

**Comment by rohskopf** [comment]
As @jmgoff mentioned, the current state of this PR is that LAMMPS needs to be compiled with the cmake flags:

```
-DPACELIB_URL= [http://github.com/jmgoff/lammps-user-pace-1/archive/refs/tags/lammps-user-pace-compute.tar.gz](https://github.com/jmgoff/lammps-user-pace-1/archive/refs/tags/lammps-user-pace-compute.tar.gz)
-DPACELIB_MD5=0616251507dffe83dbd38681d101aee8
```

This is because we are using a different `lammps-user-pace` that has been modified to support `compute pace`. 

Adding `cmake` options to pull from different repos seems like a band-aid, so hopefully we can modify the original `lammps-user-pace` so that the build will work.

**Comment by rohskopf** [comment]
If we can't change the external `lammps-user-pace` library, another option is to change the cmake file to optionally pull from our modified branch: https://github.com/lammps/lammps/blob/dc3c8da52b3ee985cac04f5b8c908d8b2dcebf0a/cmake/Modules/Packages/ML-PACE.cmake#L1

But that also seems like a band-aid.

File: examples/PACKAGES/pace/compute/README.md
Comment (stanmoore1) [inline examples/PACKAGES/pace/compute/README.md:1]: Update or remove empty file.

File: examples/test.txt
Comment (stanmoore1) [inline examples/test.txt:1]: Remove debug file.

File: src/ML-PACE/compute_pace.cpp [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+  dgradflag = utils::inumeric(FLERR, arg[5], false, lmp);
+  if (dgradflag && !bikflag)
+    error->all(FLERR,"Illegal compute pace command: dgradflag=1 requires bikflag=1");
+
+  // map[i] = which element the Ith atom type is, -1 if not mapped
+  // map[0] is not used
+  memory->create(map,ntypes+1,"pace:map");
+  /*
+  for (int i=0; i<ntypes; i++){
+    map[i+1] = i+1;
```
Comment (stanmoore1) [inline src/ML-PACE/compute_pace.cpp]: Need to remove commented out code, or make it clear why it is helpful to keep.

File: src/ML-PACE/compute_pace.cpp [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+  double cut = acecimpl -> basis_set->cutoffmax;
+  cutmax = acecimpl -> basis_set->cutoffmax;
+  double cuti;
+  double radelemall = 0.5;
+  /*
+  memory->create(cutsq,ntypes+1,ntypes+1,"pace:cutsq");
+  printf("----- looping over ntypes\n");
+  for (int i = 1; i <= ntypes; i++) {
+    printf("----- map[%d]: %d\n", i, map[i]);
+    cuti = basis_set->radial_functions->cut(map[i], map[i]);
```
Comment (stanmoore1) [inline src/ML-PACE/compute_pace.cpp]: Clean up commented code.

**Review by stanmoore1 — CHANGES_REQUESTED** [review]
Overall this looks good, but needs a little clean up as I mentioned in some of the comments.

**Review by akohlmey — APPROVED** [review]

**Comment by akohlmey** [comment]
@jmgoff I've merged with upstream and implemented requested changes (and a couple of mostly cosmetic updates, too).

Please check it out and let us know if this is now a final version or still waiting for additional modifications.

**Comment by jmgoff** [comment]
@akohlmey Thank you for making those changes and updates requested by @stanmoore1 (and for your patience with my slow updates on this)! It looks good to me and is the final version for this PR.

**Review by stanmoore1 — APPROVED** [review]

**Pull request merged**